### DISCUSSION TITLE
K8up v2

### DISCUSSION THREAD

**Discussion by ccremer** [k8up-io/k8up#236] [Ideas]
From the top of my head in no particular order:

* Redesign CRD API. We've learned a lot since the early days and should come up with a redesigned CRD API where global settings are a thing of the past. We might also make better use of core features like cronjobs or status conditions
* Get rid of our internal observer that acts as a state machine. I personally favor using K8s API as the state machine and have the operator itself stateless as much as possible. This way it has only to react upon events
* Integrate (w)restic into the K8up codebase: Schedule jobs with the same container image as the operator, but call different CLI args relevant for restores, backups etc
* Provide a developer-friendly CLI to list and restore backups, directly interacting with K8s API objects

We might make separate discussions for these ideas

**Comment by Kidswiss** [comment]
Also an idea:

Write the snapshots as CRs into the cluster. So we can provide a more k8s native way of handling restores and listing available snapshots.

**Comment by Kidswiss** [comment]
Some thoughts about the k8s native cronjobs:

I would really like to leverage them, but they have a major drawback:

They are statically defined once, with all the volumes etc. So if PVCs are added or removed we need to make sure that they reflect those changes immediately. Which in turn means that the Operator needs to observe all the PVCs in all namespaces where a schedule is deployed.
